### PR TITLE
Resolve uncovered code in transaction class

### LINKED ISSUE #76: Conditional in Database\Transaction:shouldBeginTransaction() my have an extra clause
After refactoring for PHP 7.4, the current function shouldBeginTransaction() in \Wizaplace\Etl\DatabaseTransaction is:

`return !$this->open && (0 === $this->size || 1 === $this->count);`

It's not clear to me that `0 === $this->size` should be part of that condition and in fact

`return !$this->open && 1 === $this->count;`

also passes all our current tests. Either the tests are not diagnostic or the code is inaccurate, We should determine which and resolve the inconsistency.

### PR DESCRIPTION
**Pull request by kdebisschop**
- If using transactions, any row should open a PDO transaction if one is not already open
 - If a Transaction class is destroyed, any open transactions should be closed
 - Refactor TransactionTest to provide coverage, reduce code duplication, and remove unused variables.

Resolves #76

### CODE REVIEW

File: src/Database/Transaction.php [hunk 1]
```diff
@@ -53,6 +53,14 @@ public function __construct(PDO $pdo)
         $this->pdo = $pdo;
     }
 
+    /**
+     * Code defensively by closing any open transactions when this object is destroyed.
+     */
+    public function __destruct()
+    {
+        $this->close();
```
Comment (Nicolas-Masson-Wizaplace) [inline src/Database/Transaction.php:61]: Nice idea, just a little question.

Closing the transaction will commit to the database, is it something that we want to do in case of unattended object destruction ? I mean is there case we want to rollback current transaction instead of commit ?

Comment (kdebisschop) [reply src/Database/Transaction.php:61]: I thought about that some. My logic was that if there was a way that the Transaction object could get destroyed before the pipeline completed, its behavior should be the same as if the loader was running without transactions. That is achieved by closing any pending transactions. The loader should be as resilient as possible in the sense that if the loader accepts a Row of data, we should be able to consider that data loaded.

I think if we ever found a case where we did want to roll back data that has been accepted by the loader, then it should be something we explicitly call for. In fact, as I write this, I wonder if the existing behavior to roll back transactions on error is correct. Imagine the ETL pipeline is processing a stream of data from a sensor...once the loader has accepted the data, one would think the sensor should be free to delete the data. But if it did, data could be lost not because the data itself was invalid, but because some piece of data many events later threw an exception.

In other words, do you see a reason why we should have 

```php
        try {
            call_user_func($callback);
        } catch (\Exception $exception) {
            $this->pdo->rollBack();

            throw $exception;
        }
```

and not 

```php
        try {
            call_user_func($callback);
        } catch (\Exception $exception) {
            $this->pdo->commit();

            throw $exception;
        }
```

All that being said, I am not sure there is a way an actual pipeline could destroy a Transaction in a way that did not cause larger errors. This truly is a change for defensive coding and not a response to a specific bug.

Comment (kdebisschop) [reply src/Database/Transaction.php:61]: If people agree that throwing away data that has already been loaded is also a bug, I have code that fixes it and a test that shows data is preserved from before an insert that throws an exception (in SQLite). The more I think about it the more I feel it is a bug -- the rows added by the Insert class really should be the same if you are using transactions or if you are not.

Comment (ecourtial) [reply src/Database/Transaction.php:61]: If we want to stay ACID, I think that we should not rollback automatically, but instead provide an option to enable it.

Comment (kdebisschop) [reply src/Database/Transaction.php:61]: Do you want that option to do rollback of earlier rows for this work?

It seems that could be a new feature whereas not rolling back rows that have been accepted by the Loader is just a bug fix.

I'll push the work I have now to address that bug. Then work on the option -- although I find myself hard-pressed to imagine why someone might use it.

Comment (kdebisschop) [reply src/Database/Transaction.php:61]: That makes so much sense to me that I think size 0 transactions should default to rollback and transactions larger than 0 should try to commit everything they load. I'm not even sure we need an option other than the size that we already have.

Comment (kdebisschop) [reply src/Database/Transaction.php:61]: @ecourtial I'd love your thoughts on the idea of rolling back if size=0 and committing as much as possible if size > 0

I've written it that way for now, but if we agree that is the right approach I probably also need to update the documentation.

Comment (kdebisschop) [reply src/Database/Transaction.php:61]: Oh - and as it stands right now, if we consider the issue of rolling back/committing a bug, then this can be part of 1.3 -- but I'm perfectly happy to rebase on 2.x whenever we have a 2.x branch started.

Comment (kdebisschop) [reply src/Database/Transaction.php]: I've added documentation updates to reflect that is size=0 the ETL process will be atomic, whereas if size is greater than 0 the best we can do is be durable in the sense that if a row is accepted by the insert loader it should eventually be committed.

From that perspective, I think it is fair to assume atomicity is a primary reason for wanting transactions, so I have changed the default transaction size to 0 in Insert and InsertUpdate.

Obviously, if people agree this approach makes sense, changing these defaults feeling like it is not backward compatible, even though the changes here do not require PHP 7.4+ and the v2 label is probably appropriate.

Comment (ecourtial) [reply src/Database/Transaction.php]: >Oh - and as it stands right now, if we consider the issue of rolling back/committing a bug, then this can be part of 1.3 -- but I'm perfectly happy to rebase on 2.x whenever we have a 2.x branch started.

2.x is the master branch. I created a master-1-x. Not the best option, I should have done the opposite. But that's done :/

Regarding the PR: I would like to reconsider what I said earlier. Could we imagine two possible ways, with settings, something like that:

- We insert and commit rows individually. If one fails, we continue (default behavior ?).
- The whole ETL process should be wrapped inside one transaction. If one row fails we rollback everything.

I don't know :D

Comment (kdebisschop) [reply src/Database/Transaction.php]: @ecourtial 

For your first bullet: I think that if you wanted to insert rows individually, there would be no reason to use transactions. Am I wrong? So we could make an option to either keep going or stop when encountering an error, but it would be part of Insert and InsertUpdate, not part of Transaction. So a different PR maybe?

For your second bullet: With current code in this branch, if size is zero, then all the rows are collected in one transaction and if one row fails it is all rolled back. So we already do that.

The unasked question, I think, is what makes sense if transaction size is not zero? Do we stop or keep going? I think that would just follow the same logic as your first bullet. But again, the change would be in Insert and InsertUpdate, not in Transaction. So I think that would also be a different PR than this one. Probably the same PR as the first bullet.

**Comment by ecourtial** [comment]
@kdebisschop Need a small rebase :P

**Comment by kdebisschop** [comment]
@ecourtial rebased

**Review by ecourtial — APPROVED** [review]

File: src/Database/Transaction.php (hunk 1 above)
Comment (ecourtial) [reply src/Database/Transaction.php]: @kdebisschop I agree. It is ok for me. Should we merge this PR if it is ok for you and continue any further evolution in another one as you suggest ?

Comment (kdebisschop) [reply src/Database/Transaction.php]: Yes, I think it is good to merge.

The code as currently patched has reasonable behavior, I think. If iot is one transaction, then it is either all or nothing. If it is not wrapped in a single transaction, the code tries to process as many valid inserts as it can before there is an error.

That might be the best than can be achieved.
